File: permutation/permutation.c
// C program to print all permutations with duplicates allowed
#include <stdio.h>

void display(int *arr, int length)
{
    for(int i=0; i<length; i++)
    {
        printf("%i", arr[i]);
    }

    printf("\n");
}

void swap(int *arr, int x, int y)
{
    int tmp = arr[x];
    arr[x] = arr[y];
    arr[y] = tmp;
}

void permute(int *arr, int start, int length)
{

    int i;

    if(start == length)
    {
        display(arr, length+1);
    }
    else
    {
        for(i = start; i <= length; i++)
        {
            swap(arr, start, i);

            permute(arr, start+1, length);

            swap(arr, start, i);
        }
    }
}

int main()
{
    int arr[] = {1, 2, 3};

    permute(arr, 0, 2);

    return 0;
}

File: permutation/permutation.rb
def swap(arr, x, y)
  tmp = arr[x]
  arr[x] = arr[y]
  arr[y] = tmp
end

def permute(arr, s, e)
  if s == e
    p arr
  else
    for i in s..e
      swap(arr, s, i)
      permute(arr, s+1, e)
      swap(arr, s, i)
    end
  end
end

permuate([1, 2, 3], 0, 2)
